autoscheduler/frontend/src/redux/actions/courseCards.ts

```diff
@@ -212,3 +234,71 @@ export function updateCourseCard(index: number, courseCard: CourseCardOptions, t
     return dispatch(updateCourseCardSync(index, courseCard));
   };
 }
+
+export function clearCourseCards(): ClearCourseCardsAction {
+  return { type: CLEAR_COURSE_CARDS };
+}
+
+/**
+ * Helper function to fetch updated sections for a course card and keep selected sections
+ * @param courseCard course card to update
+ * @param term term to fetch sections for
+ */
+async function updateSectionsForCourseCard(
+  courseCard: CourseCardOptions,
+  term: string,
+): Promise<CourseCardOptions> {
+  return createCourseCardFrom(courseCard, term).then((newCourseCard) => {
+    // course now updated with new sections, re-select sections from original courseCard
+    const selectedSections = new Set(courseCard.sections?.filter(({ selected }): boolean => (
+      selected
+    )).map((section): number => section.section.id));
+
+    newCourseCard.sections.forEach((section) => {
+      // eslint-disable-next-line no-param-reassign
+      if (selectedSections.has(section.section.id)) section.selected = true;
+    });
+
+    return newCourseCard;
+  });
+}
+
+/**
+ * Creates an updated course card from courseCard by making an API call for updated
+ * section information
+ * @param courseCard Course card to update
+ * @param term Term to fetch section information from
+ */
+export function replaceCourseCards(
+  courseCards: CourseCardArray,
+  term: string,
+): ThunkAction<void, RootState, undefined, CourseCardAction> {
+  // data for sections might have changed since last visit, so create a new course card
+  // for each one asynchronously and update them with data from courseCards
+  return (dispatch): void => {
+    // determine which cards actually exist
+    const cards = [];
+    for (let i = 0; i < courseCards.numCardsCreated; i++) {
+      const card = courseCards[i];
+      if (card) cards.push(card);
+    }
+
+    // create promise for each updated course card
+    const newCourseCards = cards.map((card) => (updateSectionsForCourseCard(card, term)));
+
+    // clear all course cards
+    new Promise((resolve) => { dispatch(clearCourseCards()); resolve(); }).then(async () => {
```

Comment:
Do you need to call `resolve` manually here?

Reply:
Yes, if you don't do it manually then it doesn't wait for the dispatch to complete

Reply:
In that case, I would mention that in a comment as it's not very intuitive to me

Reply:
I changed the name of the array to `courseCardPromises` after Adel's comment, do you still think the comment is necessary? Also since I refactored this code to call `.then()` on each of them I thought it would be clearer

Reply:
I think really what isn't clear to me is that you have to call `resolve()` to wait for the `dispatch(clearCourseCards())` to finish (still don't understand that tbh) - the `courseCardPromises` itself doesn't confuse me as much

Reply:
Actually I don't think it's needed, I've been testing without it and it seems to work fine. Since the action is synchronous the function should have to wait until after it finishes to continue